### PR TITLE
Device and torch._C function cleanup

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #38211 Move all torch.nn.modules type annotations inline
* #39483 Upgrade lint.
* **#38173 Device and torch._C function cleanup**

- Introduce torch.types.Device representing all "device-like" types
- Stubbed torch.device.__reduce__
- Stubbed all torch._C functions comprehensively
- Deleted _safe_call which is unused throughout the codebase

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D21497399](https://our.internmc.facebook.com/intern/diff/D21497399)

### CODE REVIEW

File: torch/_C/__init__.pyi.in [hunk 1]
```diff
@@ -17,16 +19,17 @@ T = TypeVar('T')
...
 class device:
-    type: str
-    index: _int
+    type: str  # THPDevice_type
+    index: _int  # THPDevice_index
 
+    # THPDevice_pynew
     @overload
-    def __init__(self, device: Union[_int, str]) -> None: ...
+    def __init__(self, device: Device) -> None: ...
```
Comment (rgommers) [inline torch/_C/__init__.pyi.in:27]: `Device` includes `None`, which works when `device` is a keyword in other functions, but here `torch.device(None)` actually gives a `TypeError`. So I think this change should be reverted.

File: torch/_C/__init__.pyi.in [hunk 2]
```diff
@@ -82,22 +85,46 @@ def _jit_get_operation(op_name: str) -> Callable: ...
...
+def set_num_interop_threads(nthreads: _int) -> None: ...  # THPModule_setNumInteropThreads
+def _get_cudnn_enabled() -> _bool: ...  # THPModule_userEnabledCuDNN
+def _set_cudnn_enabled(arg: _bool) -> None: ...  # THPModule_setUserEnabledCuDNN
+def _get_mkldnn_enabled() -> _bool: ...  # THPModule_userEnabledMkldnn
+def _set_mkldnn_enabled(arg: _bool) -> None: ...  # THPModule_setUserEnabledMkldnn
+def _get_cudnn_benchmark() -> _bool: ...  # THPModule_benchmarkCuDNN
+def _set_cudnn_benchmark(arg: _bool) -> None: ...  # THPModule_setBenchmarkCuDNN
+def _get_cudnn_deterministic() -> _bool: ...  # THPModule_deterministicCuDNN
+def _set_cudnn_deterministic(arg: _bool) -> None: ...  # THPModule_setDeterministicCuDNN
+# TODO: Is there a Capsule type in typing?  Doesn't seem so...
```
Comment (rgommers) [inline torch/_C/__init__.pyi.in:117]: Nope, doesn't exist. One discussion on python-dev where the suggestion is to add it to `types`, but wasn't done. From https://code.activestate.com/lists/python-dev/139675/:

>> - capsules
>> That one seems reasonable. But maybe there should just be a `types.Capsule` Type or `types.PyCapsule`, and then you can just check that the same as any other concrete type?

> But how often do you need to verify that something is a capsule, without knowing that it's the *right* capsule? At runtime, you'd usually use `PyCapsule_IsValid`, not `PyCapsule_CheckExacf`, right? So should the type checker be tracking the name too?

**Review by rgommers — COMMENTED** [review]
LGTM modulo my one comment on `device.__init__`.

clang-tidy failures can be ignored.

File: torch/csrc/Module.cpp [hunk 3]
```diff
@@ -533,7 +513,6 @@ static PyMethodDef TorchMethods[] = {
   {"_add_docstr",     (PyCFunction)THPModule_addDocStr,       METH_VARARGS, nullptr},
   {"_init_names",     (PyCFunction)THPModule_initNames,       METH_O,       nullptr},
   {"_has_distributed",(PyCFunction)THPModule_hasDistributed,  METH_NOARGS,  nullptr},
-  {"_safe_call",      (PyCFunction)(void(*)())THPModule_safeCall, METH_VARARGS | METH_KEYWORDS, nullptr},
```
Comment (albanD) [inline torch/csrc/Module.cpp:536]: This is just legacy from our old `ffi` module and we hope no one uses it outside?

**Review by albanD — APPROVED** [review]
LGTM
The corresponding cpp functions is a nice touch as it makes looking for them easy. Is mypy actually using these comments?

File: torch/csrc/Module.cpp (hunk 3 above)
Comment (ezyang) [reply torch/csrc/Module.cpp:536]: Yep

**Comment by ezyang** [comment]
> The corresponding cpp functions is a nice touch as it makes looking for them easy. Is mypy actually using these comments?

No, I just added them to make it more clear how to determine the correct annotations

**Pull request closed without merging**